lte/gateway/c/oai/tasks/amf/amf_app_ue_context_and_proc.h

```diff
@@ -0,0 +1,1001 @@
+                                //  procedure
+  ksi_t ksi;
+  uint8_t rand[AUTH_RAND_SIZE]; /* Random challenge number  */
+  uint8_t autn[AUTH_AUTN_SIZE]; /* Authentication token     */
+  imsi_t* unchecked_imsi;
+  int amf_cause;
+};
+
+typedef struct nas5g_cn_proc_s {
+  nas5g_base_proc_t base_proc;
+  cn5g_proc_type_t type;
+} nas5g_cn_proc_t;
+
+typedef struct nas5g_cn_procedure_s {
+  nas5g_cn_proc_t* proc;
+  LIST_ENTRY(nas5g_cn_procedure_s) entries;
+} nas5g_cn_procedure_t;
+
+class nas_5g_auth_info_proc_t {
+ public:
+  nas5g_cn_proc_t cn_proc;
+  success_cb_t success_notif;
+  failure_cb_t failure_notif;
+  bool request_sent;
+  uint8_t nb_vectors;
+  int nas_cause;
+  amf_ue_ngap_id_t ue_id;
+  bool resync;  // Indicates whether the authentication information is requested
+                // due to sync failure
+};
+
+nas_5g_auth_info_proc_t* nas_new_5gcn_auth_info_procedure(
+    amf_context_t* const amf_context);
+
+class m5g_authentication : public amf_context_t {
+ public:
+  int amf_proc_authentication_ksi(
+      amf_context_t* amf_context,
+      nas_amf_specific_proc_t* const amf_specific_proc, ksi_t ksi,
+      const uint8_t* const rand, const uint8_t* const autn,
+      success_cb_t success, failure_cb_t failure);
+  int amf_proc_authentication(
+      amf_context_t* amf_context,
+      nas_amf_specific_proc_t* const amf_specific_proc, success_cb_t success,
+      failure_cb_t failure);
+  int amf_proc_authentication_failure(
+      amf_ue_ngap_id_t ue_id, int amf_cause, const_bstring auts);
+  int amf_proc_authentication_complete(
+      amf_ue_ngap_id_t ue_id, AuthenticationResponseMsg* msg, int amf_cause,
+      const unsigned char* res);
+  int amf_registration_security(amf_context_t* amf_context);
+  void set_notif_callbacks_for_5g_auth_proc(nas_amf_auth_proc_t* auth_proc);
+  void set_callbacks_for_5g_auth_proc(nas_amf_auth_proc_t* auth_proc);
+  void set_callbacks_for_5g_auth_info_proc(
+      nas_5g_auth_info_proc_t* auth_info_proc);
+  int amf_send_authentication_request(
+      amf_context_t* amf_context, nas_amf_auth_proc_t* auth_proc);
+};
+
+class nas_proc {
+ public:
+  int nas_proc_establish_ind(
+      const amf_ue_ngap_id_t ue_id, const bool is_mm_ctx_new,
+      const tai_t originating_tai, const ecgi_t ecgi,
+      const m5g_rrc_establishment_cause_t as_cause, const s_tmsi_m5_t s_tmsi,
+      bstring msg);
+  nas_amf_registration_proc_t* get_nas_specific_procedure_registration(
+      const amf_context_t* ctxt);
+  bool is_nas_specific_procedure_registration_running(
+      const amf_context_t* ctxt);
+  int nas5g_message_decode(
+      unsigned char* buffer, amf_nas_message_t* nas_msg, int length,
+      amf_security_context_t* amf_security_context,
+      amf_nas_message_decode_status_t* decode_status);
+  nas_amf_common_proc_t* get_nas5g_common_procedure(
+      const amf_context_t* const ctxt, amf_common_proc_type_t proc_type);
+};
```

Comment:
does not make much sense to me why all these are in a class...

Reply:
fixed